src/zcash/JoinSplit.cpp

```diff
@@ -0,0 +1,344 @@
+        }
+    }
+    void loadVerifyingKey(std::string path) {
+        loadFromFile(path, vk);
+    }
+    void saveVerifyingKey(std::string path) {
+        if (vk) {
+            saveToFile(path, *vk);
+        } else {
+            throw std::runtime_error("cannot save verifying key; key doesn't exist");
+        }
+    }
+
+    void generate() {
+        protoboard<FieldT> pb;
+
+        joinsplit_gadget<FieldT, NumInputs, NumOutputs> g(pb);
+        g.generate_r1cs_constraints();
+
+        const r1cs_constraint_system<FieldT> constraint_system = pb.get_constraint_system();
+        r1cs_ppzksnark_keypair<ppzksnark_ppT> keypair = r1cs_ppzksnark_generator<ppzksnark_ppT>(constraint_system);
+
+        pk = keypair.pk;
+        vk = keypair.vk;
+    }
+
+    JoinSplitCircuit() {}
+
+    bool verify(
+        const std::string& proof,
+        const uint256& pubKeyHash,
+        const uint256& randomSeed,
+        const boost::array<uint256, NumInputs>& hmacs,
+        const boost::array<uint256, NumInputs>& nullifiers,
+        const boost::array<uint256, NumOutputs>& commitments,
+        uint64_t vpub_old,
+        uint64_t vpub_new,
+        const uint256& rt
+    ) {
+        if (!vk) {
+            throw std::runtime_error("JoinSplit verifying key not loaded");
+        }
+
+        r1cs_ppzksnark_proof<ppzksnark_ppT> r1cs_proof;
+        std::stringstream ss;
+        ss.str(proof);
+        ss >> r1cs_proof;
+
+        uint256 h_sig = this->h_sig(randomSeed, nullifiers, pubKeyHash);
+
+        auto witness = joinsplit_gadget<FieldT, NumInputs, NumOutputs>::witness_map(
+            rt,
+            h_sig,
+            hmacs,
+            nullifiers,
+            commitments,
+            vpub_old,
+            vpub_new
+        );
+
+        return r1cs_ppzksnark_verifier_strong_IC<ppzksnark_ppT>(*vk, witness, r1cs_proof);
+    }
+
+    std::string prove(
+        const boost::array<JSInput, NumInputs>& inputs,
+        const boost::array<JSOutput, NumOutputs>& outputs,
+        boost::array<Note, NumOutputs>& output_notes,
+        boost::array<ZCNoteEncryption::Ciphertext, NumOutputs>& ciphertexts,
+        uint256& ephemeralKey,
+        const uint256& pubKeyHash,
+        uint256& randomSeed,
+        boost::array<uint256, NumInputs>& hmacs,
+        boost::array<uint256, NumInputs>& nullifiers,
+        boost::array<uint256, NumOutputs>& commitments,
+        uint64_t vpub_old,
+        uint64_t vpub_new,
+        const uint256& rt
+    ) {
+        if (!pk) {
+            throw std::runtime_error("JoinSplit proving key not loaded");
+        }
+
+        // Compute nullifiers of inputs
+        for (size_t i = 0; i < NumInputs; i++) {
+            nullifiers[i] = inputs[i].nullifier();
+        }
+
+        // Sample randomSeed
+        randomSeed = random_uint256();
+
+        // Compute h_sig
+        uint256 h_sig = this->h_sig(randomSeed, nullifiers, pubKeyHash);
```

Comment:
Where is ``pubKeyHash`` computed? I don't see the string "ZcashECDSAPubKey" (or equivalent character array) anywhere.